src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/UserAccessScopeMapper.kt

```diff
@@ -0,0 +1,51 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.authorization
+
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.AuthUser
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.DynamicFrameworkContract
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ServiceProvider
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.DynamicFrameworkContractRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ServiceProviderRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.HMPPSAuthService
+import java.lang.RuntimeException
+
+data class UserAccessScope(
+  val serviceProvider: ServiceProvider,
+  val contracts: List<DynamicFrameworkContract>,
+)
+
+@Component
+class UserAccessScopeMapper(
+  private val hmppsAuthService: HMPPSAuthService,
+  private val serviceProviderRepository: ServiceProviderRepository,
+  private val dynamicFrameworkContractRepository: DynamicFrameworkContractRepository,
+) {
+  private val serviceProviderGroupPrefix = "INT_SP_"
+  private val contractGroupPrefix = "INT_CG_"
+
+  // get the user access scope for a given service provider user
+  fun forServiceProviderUser(user: AuthUser): UserAccessScope {
+    val groups = hmppsAuthService.getUserGroups(user)
+      ?: throw RuntimeException("pp or non existent user")
```

Comment:
these exceptions are not ideal - i think we should write some new exception classes and handlers for auth errors. similarly for the `ServerWebInputExceptions` below

Reply:
In my view, this is data at this point, becoming an exception one level up

We could use the output of this function as a way to display the problems with the user:
```
scope = forServiceProviderUser(user)
scope.hasProblems() -> throw AuthException
```
where `scope.problems` could be
```
[
  "user is not a provider user",
  "provider with id WEIRD does not exist",
  "contract with reference WOO does not exist",
  "contract C2 belongs to a different provider"
]
```
etc

Treating problems as data would allow us to test all scenarios and not worry about the shortcut paths. (And possibly to serve content even if there's a minor misconfiguration)

Reply:
ok i've put together something along these lines. 

to stop the caller of this from explicitly having to check for config errors (easily forgotten), i make this class raise an exception, but with all possible configuration errors included.